pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,345 @@
+            Allowed names are 'PauliX', 'PauliY', 'PauliZ' and 'Hadamard'. If no list is provided,
+            the function will diagonalize everything.
+
+    Returns:
+        diagonalizing_gates: A list of operations to be applied to diagonalize the observable
+        new_obs: the relevant measurement to perform after applying diagonalzing_gates to get the
+            correct measurement output
+        _visited_obs: an up-to-date record of the observables and wires
+            encountered on the tape so far
+    """
+
+    if _visited_obs is None:
+        _visited_obs = ([], [])
+
+    if not isinstance(observable, (qml.X, qml.Y, qml.Z, qml.Hadamard, qml.Identity)):
+        return _diagonalize_compound_observable(
+            observable, _visited_obs, supported_base_obs=supported_base_obs
+        )
+
+    # also validates that the wire hasn't already been diagonalized and updated _visited_obs
+    switch_basis = observable.name not in supported_base_obs
+    diagonalize, _visited_obs = _check_if_diagonalizing(observable, _visited_obs, switch_basis)
+
+    if isinstance(observable, qml.Z):  # maybe kind of redundant
+        return [], observable, _visited_obs
+
+    new_obs = qml.Z(wires=observable.wires) if switch_basis else observable
+    diagonalizing_gates = observable.diagonalizing_gates() if diagonalize else []
+
+    return diagonalizing_gates, new_obs, _visited_obs
+
+
+def _get_obs_and_gates(obs_list, _visited_obs, supported_base_obs=_default_supported_obs):
+    """Calls _diagonalize_observable on each observable in a list, and returns the full result
+    for the list of observables"""
+
+    new_obs = []
+    diagonalizing_gates = []
+
+    for o in obs_list:
+        gates, obs, _visited_obs = _diagonalize_observable(o, _visited_obs, supported_base_obs)
+        if gates:
+            diagonalizing_gates.extend(gates)
+        new_obs.append(obs)
+
+    return diagonalizing_gates, new_obs, _visited_obs
+
+
+@singledispatch
+def _diagonalize_compound_observable(
+    observable, _visited_obs, supported_base_obs=_default_supported_obs
+):
+    """Takes an observable consisting of multiple other observables, and changes all
+    unsupported obs to the measurement basis. Applies diagonalizing gates if changing
+    the basis of an observable whose diagonalizing gates have not already been applied."""
+
+    raise NotImplementedError(
+        f"Unable to convert observable of type {type(observable)} to the measurement basis"
+    )
+
+
+@_diagonalize_compound_observable.register
+def _diagonalize_symbolic_op(
+    observable: SymbolicOp, _visited_obs, supported_base_obs=_default_supported_obs
+):
+    diagonalizing_gates, new_base, _visited_obs = _diagonalize_observable(
+        observable.base, _visited_obs, supported_base_obs
+    )
+
+    params, hyperparams = observable.parameters, observable.hyperparameters
+    hyperparams["base"] = new_base
```

Comment:
```suggestion
    params, hyperparams = observable.parameters, observable.hyperparameters
    hyperparams = copy.copy(hyperparams)
    hyperparams["base"] = new_base
```
since dictionaries are mutable, we should be careful about not updating it in place.